### PR TITLE
[WIP] Script generating a changelog

### PR DESCRIPTION
**Pull request by tinygrasshopper**
WIP generate changelog

feat: markdown generator

mark endpoints as added

feat: refactor code

update start date to 2024-04-25

- also filter out /openapi changes

### CODE REVIEW

File: docs/snyk-api/CHANGELOG.md
Comment (tinygrasshopper) [inline docs/snyk-api/CHANGELOG.md:1]: can we exclude this file from the diff; content changes can be generated and merged separately with approval from the docs team

File: tools/api-docs-generator/changelog/changelog.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+	for _, endpointChange := range changes {
+		if strings.HasPrefix(endpointChange.Path, "/openapi") {
+			continue
+		}
+		filteredChanges = append(filteredChanges, endpointChange)
+	}
+	return filteredChanges
+}
+
+func WriteToChangeLog(markdown *md.Markdown, groupedChanges []ChangesByEndpoint, baseVersion, nextVersionURL, lastSyncVersion string) error {
```
Comment (tinygrasshopper) [inline tools/api-docs-generator/changelog/changelog.go:152]:
```suggestion
func writeToChangeLog(markdown *md.Markdown, groupedChanges []ChangesByEndpoint, baseVersion, nextVersionURL, lastSyncVersion string) error {
```
This function could be a package level one

File: docs/snyk-api/CHANGELOG.md
Comment (tinygrasshopper) [inline docs/snyk-api/CHANGELOG.md:1]: can we exclude this file from the change-set; lets get the content changes reviewed separately 

Have a PR here: https://github.com/snyk/user-docs/pull/301

File: tools/api-docs-generator/main.go [hunk 2]
```diff
@@ -17,12 +20,32 @@ func main() {
...
+		log.Panic(err)
+	}
+
+	if updatedToVersion != "" {
+		syncStateCfg.LastSyncedVersion = updatedToVersion
+		updateSyncStateErr := config.UpdateSyncState(path.Join(docsDirectory, cfg.Changelog.SyncStateFile), syncStateCfg)
+		if err != nil {
+			log.Panic(updateSyncStateErr)
+		}
+	}
```
Comment (tinygrasshopper) [inline tools/api-docs-generator/main.go:46]: IMO reading the sync state and writing it back should be done in the `UpdateChangelog` function; internals of `UpdateChangelog` are leaking out here

File: tools/api-docs-generator/versions/versions.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	}
+	err = resp.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return versions, nil
+}
+
+func Get(ctx context.Context, urlToGet string) (*http.Response, error) {
```
Comment (tinygrasshopper) [inline tools/api-docs-generator/versions/versions.go:34]: This should be a private function to the package

File: tools/api-docs-generator/changelog/changelog.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+			markdown.H3f("%s - `%s` - Added", changeGroup.Operation, changeGroup.Path)
+			parsedURL, err := url.Parse(nextVersionURL)
+			if err != nil {
+				return err
+			}
+			document, err := loader.LoadFromURI(parsedURL)
+			if err != nil {
+				return err
+			}
+			markdown.BulletList(versions.ReplaceWithCodeQuotes(document.Paths.Find(changeGroup.Path).Operations()[changeGroup.Operation].Description))
```
Comment (tinygrasshopper) [inline tools/api-docs-generator/changelog/changelog.go:184]: `versions.ReplaceWithCodeQuotes` does not really sound right; the `ReplaceWithCodeQuotes` function should be moved to the changelog package

File: tools/api-docs-generator/cmd/historical-change-log/main.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	"context"
+	"log"
+	"time"
+
+	"github.com/snyk/user-docs/tools/api-docs-generator/changelog"
+
+	"github.com/snyk/user-docs/tools/api-docs-generator/config"
+)
+
+const configFile = "tools/api-docs-generator/config.yml"
```
Comment (tinygrasshopper) [inline tools/api-docs-generator/cmd/historical-change-log/main.go:13]: `configFile` value should come from a param; currently it assumes the location from which the code is run; you can use `os.Args[1]` to get positional arguments

File: docs/snyk-api/CHANGELOG.md
Comment (corneliu-petrescu) [reply docs/snyk-api/CHANGELOG.md:1]: so, I should simply delete the file?

**Comment by tinygrasshopper** [comment]
Also, can we add a make helper task to run the historical log; like the `make run` task

File: docs/snyk-api/CHANGELOG.md
Comment (tinygrasshopper) [reply docs/snyk-api/CHANGELOG.md:1]: if you copy the current contents of main branch onto this branch; it wont showup on the PR anymore

Comment (tinygrasshopper) [reply docs/snyk-api/CHANGELOG.md:1]: if you copy the current contents of main branch onto this branch; it wont showup on the PR anymore

File: tools/api-docs-generator/versions/versions.go (hunk 3 above)
Comment (corneliu-petrescu) [reply tools/api-docs-generator/versions/versions.go:34]: This is reused in spec_fetcher. I can't make it private

**Pull request closed without merging**